src/lib/CoreInstance.cc

```diff
@@ -0,0 +1,304 @@
+  } else if (setInstructionMemory_ && instructionMemory_ == nullptr) {
+    std::cerr << "Instruction memory not set. External instruction memory "
+                 "interface must be manually set using the "
+                 "setL1InstructionMemory(...) function."
+              << std::endl;
+    exit(1);
+  }
+
+  // Construct architecture object
+  arch_ =
+      std::make_unique<simeng::arch::aarch64::Architecture>(kernel_, config_);
+
+  // Construct branch predictor object
+  predictor_ = std::make_unique<simeng::GenericPredictor>(config_);
+
+  // Extract port arrangement from config file
+  auto config_ports = config_["Ports"];
+  std::vector<std::vector<uint16_t>> portArrangement(config_ports.size());
+  for (size_t i = 0; i < config_ports.size(); i++) {
+    auto config_groups = config_ports[i]["Instruction-Group-Support"];
+    // Read groups in associated port
+    for (size_t j = 0; j < config_groups.size(); j++) {
+      portArrangement[i].push_back(config_groups[j].as<uint16_t>());
+    }
+  }
+  portAllocator_ = std::make_unique<simeng::pipeline::BalancedPortAllocator>(
+      portArrangement);
+
+  // Configure reservation station arrangment
+  std::vector<std::pair<uint8_t, uint64_t>> rsArrangement;
+  for (size_t i = 0; i < config_["Reservation-Stations"].size(); i++) {
+    // Iterate over each reservation station in config
+    auto reservation_station = config_["Reservation-Stations"][i];
+    for (size_t j = 0; j < reservation_station["Ports"].size(); j++) {
+      // Iterate over issue ports in reservation station
+      uint8_t port = reservation_station["Ports"][j].as<uint8_t>();
+      if (rsArrangement.size() < port + 1) {
+        // Resize vector to match number of execution ports available across
+        // all reservation stations
+        rsArrangement.resize(port + 1);
+      }
+      // Map an execution port to a reservation station
+      rsArrangement[port] = {i, reservation_station["Size"].as<uint16_t>()};
+    }
+  }
+
+  // Construct the core object based on the defined simulation mode
+  uint64_t entryPoint = process_->getEntryPoint();
+  if (mode_ == SimulationMode::Emulation) {
+    core_ = std::make_shared<simeng::models::emulation::Core>(
+        *instructionMemory_, *dataMemory_, entryPoint, processMemorySize_,
+        *arch_);
+  } else if (mode_ == SimulationMode::InOrderPipelined) {
+    core_ = std::make_shared<simeng::models::inorder::Core>(
+        *instructionMemory_, *dataMemory_, processMemorySize_, entryPoint,
+        *arch_, *predictor_);
+  } else if (mode_ == SimulationMode::OutOfOrder) {
+    core_ = std::make_shared<simeng::models::outoforder::Core>(
+        *instructionMemory_, *dataMemory_, processMemorySize_, entryPoint,
+        *arch_, *predictor_, *portAllocator_, rsArrangement, config_);
+  }
+
+  createSpecialFileDirectory();
+
+  return;
+}
+
+void CoreInstance::createSpecialFileDirectory() {
+  simeng::SpecialFileDirGen SFdir = simeng::SpecialFileDirGen(config_);
```

Comment:
This can be done inside the if clause - no point constructing an object that may not be used